### PR TITLE
ENH add script to mangle homebrew

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [ ] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ocefpaf** [comment]
Some feedstocks need quartz from Homebrew. It would be nice to support "keeping it" in those cases.

**Comment by beckermr** [comment]
@ocefpaf How does this work now? The CI scripts fully remove homebrew no matter what right now.

**Comment by beckermr** [comment]
OK @conda-forge/core this script works. I tested it here: https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/43

ready for review

**Comment by beckermr** [comment]
We will need a PR to smithy to enable it.

**Comment by ocefpaf** [comment]
> @ocefpaf How does this work now? The CI scripts fully remove homebrew no matter what right now.

By hacking the CI config after a re-render and installing Xquartz before the removal. This script will break those feedstocks. Our original solution was already quite bad but due to the limited amount of feedstocks that needed this we never bother to do something like yum_requirements.txt.

**Comment by isuruf** [comment]
@beckermr, if this is moving individual files, then it's no different from deleting.

**Comment by beckermr** [comment]
🤦 

Moving directories is not really an option since that might move something we need IIUIC. I can OFC code that up, but unclear if that is a good idea. Closing.

**Comment by beckermr** [comment]
@isuruf this is a filesystem dependent statement I think. I have one more idea here

**Comment by isuruf** [comment]
Do you have a list of files in `/usr/local`? I think we can just delete `/usr/local`.

**Comment by beckermr** [comment]
I am getting the list of files from the homebrew uninstall script using --dry-run. 

Most of them are on common paths, so I am removing those paths first. Then I go through and clean up anything else that is left.

**Comment by beckermr** [comment]
This seems to work well

```
MOVED /usr/local/Caskroom -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/hee1d3c_Caskroom_hee1d3c
MOVED /usr/local/Cellar -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h8c86b1_Cellar_h8c86b1
MOVED /usr/local/Homebrew -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h3fbf8d_Homebrew_h3fbf8d
MOVED /usr/local/lib/gcc -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h147faa_gcc_h147faa
MOVED /usr/local/lib/perl5 -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h56f385_perl5_h56f385
MOVED /usr/local/include/c++ -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/hc5a118_c++_hc5a118
MOVED /usr/local/Frameworks/Python.framework -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h948cc6_Python.framework_h948cc6
MOVED /usr/local/share/info/dir -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/hb3bbe5_dir_hb3bbe5
MOVED /usr/local/lib/node_modules/appcenter-cli/node_modules/extglob/lib/.DS_Store -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/hd1df6b_.DS_Store_hd1df6b
MOVED /usr/local/lib/node_modules/appcenter-cli/node_modules/micromatch/lib/.DS_Store -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/hfd8ffc_.DS_Store_hfd8ffc
MOVED /usr/local/lib/ruby/gems/2.6.0/gems/fastlane-2.144.0/spaceship/lib/spaceship/connect_api/.DS_Store -> /usr/local/mangled_homebrew_files_18fa6b9406364b6f97853cbf870f7ddf/h399838_.DS_Store_h399838

```

File: recipe/conda_forge_ci_setup/mangle_homebrew.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        else:
+            parts = [line]
+
+        for p in parts:
+            # ignore homebrew printing stuff
+            if p.startswith("==>"):
+                continue
+
+            # sometimes it does this
+            if p.startswith("Would delete "):
```
Comment (mbargull) [inline recipe/conda_forge_ci_setup/mangle_homebrew.py:116]: This does not catch https://github.com/Homebrew/install/blob/db352bc12ceee10047f1da9314f624390b576697/uninstall#L244 .

Comment (beckermr) [reply recipe/conda_forge_ci_setup/mangle_homebrew.py:116]: woops the space

Comment (beckermr) [reply recipe/conda_forge_ci_setup/mangle_homebrew.py]: fixed and good catch!

**Comment by mbargull** [comment]
Is this really worth it? Parsing the output of some (non-versioned) script seems really brittle.
How often do those 10 minute runs happen? IIRC 1-2 minutes are what we usually get, right?

**Comment by beckermr** [comment]
I have the same thoughts here @mbargull. Plus if the problem is the homebrew script itself, then this won't help.

File: recipe/conda_forge_ci_setup/mangle_homebrew.py [hunk 2]
```diff
@@ -113,8 +113,8 @@ def main():
                 continue
 
             # sometimes it does this
-            if p.startswith("Would delete "):
-                p = p[len("Would delete "):]
+            if p.startswith("Would delete"):
+                p = p[len("Would delete"):]
```
Comment (mbargull) [inline recipe/conda_forge_ci_setup/mangle_homebrew.py:117]: still needs to remove the leading `:`.

**Comment by beckermr** [comment]
IDK what we usually get.

**Comment by mbargull** [comment]
Ca 1:30 was something I think we had, but that's just from memory (and nothing I often look at, ofc >.<).

**Comment by beckermr** [comment]
Looks like 2:30 seconds here: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=137838&view=logs&jobId=4b3757f9-99c5-5891-6c0c-93d7030cdf7d&j=4b3757f9-99c5-5891-6c0c-93d7030cdf7d&t=35ac9962-a14c-5f54-e430-fcd4c09e4a5e

**Comment by mbargull** [comment]
Looking at the different runs in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=137838&view=results , it looks like the actual recipe building is more time consistent than the Homebrew removal... well, that's annoying.

**Comment by beckermr** [comment]
That recipe is a shell containing nothing and not typical.

**Comment by mbargull** [comment]
Sure, I just mean even `conda-build`+`conda` seem to be less erratic than Homebrew (or likely imposed I/O from it) in that case.

**Comment by beckermr** [comment]
Yup it is really crazy! :/

**Comment by isuruf** [comment]
If all homebrew files are under `/usr/local`, we should move everything inside `/usr/local` except `/usr/local/miniconda`.
There are other software like nodejs, powershell and other stuff installed there that we don't need.

**Comment by beckermr** [comment]
OK @isuruf - we can try that. There are some caches and logs that are not in /usr/local. 

I left bin intact just in case, but I can remove that too.

**Comment by isuruf** [comment]
Here's a simple bash script
https://github.com/conda-forge/_azure_pipelines_test/blob/master/azure-pipelines.yml#L42-L45

**Comment by beckermr** [comment]
For sure. I thought maybe moving the code into here from smithy might help if we need to push bug fixes more quickly

**Comment by isuruf** [comment]
Yes, we can have a bash script here instead of a python script.

**Comment by beckermr** [comment]
I am agnostic as to bash or python, so if you'd like to use your version then, by all means, I can push it. I think the python entrypoint one is fixed now. I made a couple of stupid mistakes.

File: recipe/conda_forge_ci_setup/mangle_homebrew.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def main():
+    # make the mangled path
+    mangled_dir = "/usr/local/mangled_homebrew_files_%s" % uuid.uuid4().hex
+    os.makedirs(mangled_dir, exist_ok=True)
+
+    excluded_dirs = [
+        os.path.basename(mangled_dir),
+        "bin",
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/mangle_homebrew.py:43]: Why keep `bin`?

Comment (beckermr) [reply recipe/conda_forge_ci_setup/mangle_homebrew.py:43]: I was unsure if it had things we needed, but I can remove it.

File: recipe/conda_forge_ci_setup/mangle_homebrew.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    # move all of the stuff except miniconda
+    potential_dirs = os.listdir("/usr/local")
+    for _pth in potential_dirs:
+        pth = os.path.join("/usr", "local", _pth)
+        if (
+            os.path.exists(pth)
+            and os.path.isdir(pth)
+            and _pth not in excluded_dirs
+        ):
+            mangled_pth = _mangele_path(pth, mangled_dir)
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/mangle_homebrew.py:56]: Let's just use `/usr/local/conda_mangled` and keep this script simple.

**Comment by beckermr** [comment]
Thanks for the fixes @isuruf!

**Comment by beckermr** [comment]
I'd be nice to keep the debugging statements so that ppl can see what happened if they hit issues.

**Comment by isuruf** [comment]
@ocefpaf, this script will replace the script at https://github.com/conda-forge/numpy-feedstock/blob/master/.azure-pipelines/azure-pipelines-osx.yml#L33-L38 so you can still comment that out.

**Pull request merged**